### PR TITLE
Fixes for item removal

### PR DESCRIPTION
**Pull request by Slayer95**
Neither Fling nor Natural Gift triggered Unburden.
Mails removed by Knock Off were recyclable and did not trigger Unburden either.
Flinged Air Balloons and Eject Buttons were unaffected by Pick Up.
Implemented function pokemon.tossItem, for removing items that would be affected by Recycle and Pick Up, while detecting the removal and triggering a TossItem event.
Replaced the boolean pokemon.usedItemThisTurn by the object pokemon.lastItemTurnData to track the way in which the item was removed.

### CODE REVIEW

**Comment by ascriptmaster** [comment]
is this done or not?

**Comment by Slayer95** [comment]
No

**Comment by Joimer** [comment]
Happy birthday, pull request! What's up with this, what's the current state on item removal mechanics?

File: data/items.js [hunk 1]
```diff
@@ -2376,7 +2376,9 @@ exports.BattleItems = {
 		id: "mail",
 		name: "Mail",
 		spritenum: 403,
-		onTakeItem: false,
+		onTakeItem: function(item, source) {
+			if (!source.volatiles['knockoff']) return false;
```
Comment (ascriptmaster) [inline data/items.js]: It's probably better to check the battle's `activeMove` instead of pinning a volatile on Knock Off just for the purposes of Mail.

**Comment by ascriptmaster** [comment]
Also while we're at it, Pickup should be able to pick up items from adjacent allies, but our implementation only picks up items from opponents.

**Comment by ascriptmaster** [comment]
Making a note of this conversation from IRC here as well:
> 11:23	Marty:	that's something else I have to do once that pull request is accepted or whatever - fix Knock Off
> 11:23	Ascriptmaster:	fix Knock Off to do what
> 11:24	Marty:	Knock Off doesn't actually reset the last used item, it's Recycle and Harvest that do it
> 11:24	Ascriptmaster:	what do you mean?
> 11:25	Marty:	I mean when you Knock Off an item, that Pokemon can still Recycle an item used before that, assuming it never used an item since
> 11:25	Marty:	so like, I use a Sitrus Berry then you Trick me something and knock it off; I can still Recycle the Sitrus Berry
> 11:26	Marty:	but if I Recycle the Sitrus Berry and you knock it off, I can't get it back because Recycle literally gave me the Sitrus Berry that was sitting in my last used item slot

File: data/items.js (hunk 1 above)
Comment (Zarel) [inline data/items.js]: I agree with @ascriptmaster here.

**Comment by Slayer95** [comment]
Superseeded by https://github.com/Zarel/Pokemon-Showdown/pull/1560

**Pull request closed without merging**